airbyte-commons-worker/src/main/java/io/airbyte/workers/internal/VersionedAirbyteStreamFactory.java

```diff
@@ -43,6 +56,79 @@ public VersionedAirbyteStreamFactory(final AirbyteMessageSerDeProvider serDeProv
                                        final MdcScope.Builder containerLogMdcBuilder) {
     // TODO AirbyteProtocolPredicate needs to be updated to be protocol version aware
     super(new AirbyteProtocolPredicate(), LOGGER, containerLogMdcBuilder);
+    Preconditions.checkNotNull(protocolVersion);
+    this.serDeProvider = serDeProvider;
+    this.migratorFactory = migratorFactory;
+    this.initializeForProtocolVersion(protocolVersion);
+  }
+
+  /**
+   * Create the AirbyteMessage stream.
+   *
+   * If detectVersion is set to true, it will decide which protocol version to use from the content of
+   * the stream rather than the one passed from the constructor.
+   */
+  @SneakyThrows
+  @Override
+  public Stream<AirbyteMessage> create(final BufferedReader bufferedReader) {
+    if (shouldDetectVersion) {
+      final Optional<Version> versionMaybe = detectVersion(bufferedReader);
+      if (versionMaybe.isPresent()) {
+        logger.info("Detected Protocol Version {}", versionMaybe.get().serialize());
+        initializeForProtocolVersion(versionMaybe.get());
+      } else {
+        // No version found, use the default as a fallback
+        logger.info("Unable to detect Protocol Version, assuming protocol version {}", fallbackVersion.serialize());
+        initializeForProtocolVersion(fallbackVersion);
+      }
+    }
+    return super.create(bufferedReader);
+  }
+
+  /**
+   * Attempt to detect the version by scanning the stream
+   *
+   * Using the BufferedReader reset/mark feature to get a look-ahead. We will attempt to find the
+   * first SPEC message and decide on a protocol version from this message.
+   *
+   * @param bufferedReader the stream to read
+   * @return The Version if found
+   * @throws IOException
+   */
+  private Optional<Version> detectVersion(final BufferedReader bufferedReader) throws IOException {
+    bufferedReader.mark(BUFFER_READ_AHEAD_LIMIT);
+    // Cap detection to the first 10 messages
```

Comment:
I don't understand why we cap at 10.

Reply:
I should probably add more details in the comments, ideally, the version detection should only read the first message.
In practice, the connectors are sending log messages before the actual `SPEC` message. The main idea behind capping is that we should not be reading ahead too far in order to detect the protocol version.
Another constraint is how `BufferedReader::mark/reset` work, if we read ahead too far, it will throw and exception and we won't be able to reset.

Reply:
Added more comments.